[SE-0358][stdlib] Adopt primary associated types

lorentey opened the pull request:
This draft PR experiments with the adoption of the proposed new language feature for [lightweight same-type syntax for primary associated types][SE-0346] in the Swift Standard Library.

[SE-0346]: https://github.com/apple/swift-evolution/blob/main/proposals/0346-light-weight-same-type-syntax.md

slavapestov commented on stdlib/public/core/SIMDVector.swift:
```diff
@@ -71,11 +73,17 @@ public protocol SIMDScalar {
+public protocol SIMD<
+  // FIXME: This ought to refer to `SIMDStorage.Scalar`, not introduce a new
+  // associated type.
+  Scalar
```
This is an ABI break

lorentey commented:
This now reflects the latest revision of SE-0346, which made things a lot nicer!

lorentey committed:
[stdlib] Don’t mark Key as primary on Keyed*ContainerProtocol

benrimmington commented on stdlib/public/core/Codable.swift:
```diff
@@ -938,7 +938,7 @@ public struct KeyedEncodingContainer<K: CodingKey> :
 /// Decoders should provide types conforming to `UnkeyedDecodingContainer` for
 /// their format.
-public protocol KeyedDecodingContainerProtocol {
+public protocol KeyedDecodingContainerProtocol<Key> {
```
This was missed in the previous commit:

```suggestion
public protocol KeyedDecodingContainerProtocol {
```

lorentey committed:
Update stdlib/public/core/Codable.swift

Co-authored-by: Ben Rimmington <me@benrimmington.com>

lorentey commented:
@swift-ci test

slavapestov commented on stdlib/public/core/LazySequence.swift:
```diff
@@ -128,7 +128,8 @@
 /// effects.
-public protocol LazySequenceProtocol: Sequence {
+public protocol LazySequenceProtocol<Elements>: Sequence
+where Elements.Element == Element {
```
Did you mean to add this? The constraint is already on the associatedtype below

lorentey replied:
🤦‍♂️ No, this was left over from the previous version.

lorentey committed:
[stdlib] LazySequenceProtocol: Take out duplicate where clause

lorentey commented:
(The failed builds are expected; this feature is currently gated on `-enable-parameterized-protocol-types`, which this PR does not enable for the stdlib yet.)

lorentey commented:
@swift-ci test

milseman commented:
Status of this work? This is blocking regex algorithms, which currently have an ABI of returning Array (i.e. force an eager and allocation)

lorentey committed:
[stdlib] Adopt primary associated types in the stdlib

lorentey committed:
[stdlib] Adopt primary associated types in _Concurrency

lorentey committed:
[stdlib] Take out primary associated types on LazySequenceProtocol/LazyCollectionProtocol for now

lorentey committed:
[stdlib][Distributed] Add primary associated types for distributed actor protocols

lorentey committed:
Revert "[stdlib][Distributed] Add primary associated types for distributed actor protocols"

This reverts commit eed5bd131a6deccf3827319532e11a5b0e9e034f.

lorentey commented:
@swift-ci test

lorentey committed:
[stdlib] OptionSet: Remove the primary associated type declaration

`Element` is expected to always be `Self`. `RawValue` would be a better choice for the primary type ("option set of UInt32"), but to avoid confusion, it seems better to just omit the declaration, at least for now.
